apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerDataReader.cs

```diff
@@ -109,130 +113,202 @@ public override long GetBytes(int i, long fieldOffset, byte[] buffer, int buffer
+#endif
         }
 
         /// <inheritdoc />
         public override System.Type GetFieldType(int i)
         {
-            throw new NotImplementedException();
+            var fieldMetadata = GetMetadataAsync(CancellationToken.None).Result.RowType.Fields[i];
+            return fieldMetadata.Type.Code.GetDefaultClrTypeFromSpannerType();
+        }
+
+        private V1.Type GetSpannerFieldType(int i)
+        {
+            var fieldMetadata = GetMetadataAsync(CancellationToken.None).Result.RowType.Fields[i];
+            return fieldMetadata.Type;
         }
 
         /// <inheritdoc />
         public override float GetFloat(int i)
         {
-            throw new NotImplementedException();
+            return (float)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(float));
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="i"></param>
+        /// <returns></returns>
+        public Value GetJsonValue(int i)
+        {
+            return (Value)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(Value));
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="i"></param>
+        /// <returns></returns>
+        public Timestamp GetTimestamp(int i)
+        {
+            //TODO, figure out how to better convert to Timestamp from RFC 3339
+            return Timestamp.FromDateTime((DateTime)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(DateTime)));
         }
 
         /// <inheritdoc />
         public override Guid GetGuid(int i)
         {
-            throw new NotImplementedException();
+            return (Guid)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(Guid));
         }
 
         /// <inheritdoc />
         public override short GetInt16(int i)
         {
-            throw new NotImplementedException();
+            return (short)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(short));
         }
 
         /// <inheritdoc />
         public override int GetInt32(int i)
         {
-            throw new NotImplementedException();
+            return (int)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(int));
         }
 
         /// <inheritdoc />
         public override long GetInt64(int i)
         {
-            throw new NotImplementedException();
+            return (long)_innerList[i].ConvertToClrType(GetSpannerFieldType(i), typeof(long));
         }
 
         /// <inheritdoc />
         public override string GetName(int i)
         {
-            throw new NotImplementedException();
+            return _resultset.GetMetadataAsync(CancellationToken.None).Result.RowType.Fields[i].Name;
         }
 
         /// <inheritdoc />
         public override int GetOrdinal(string name)
         {
-            throw new NotImplementedException();
+            var fields = _resultset.GetMetadataAsync(CancellationToken.None).Result.RowType.Fields;
```

Comment:
Do we need to get the metadata every time? The `GetMetadataAsync()` method below makes me think not?